lib/streamlit/runtime/state/query_params.py

```diff
@@ -0,0 +1,124 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import Any, Dict, Iterator, List, MutableMapping, Union
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+
+@dataclass
+class QueryParams(MutableMapping[str, Any]):
+    """A lightweight wrapper of a dict that sends forwardMsgs when state changes.
+    It stores str keys with str and List[str] values.
+    """
+
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __init__(self):
+        # avoid using ._query_params as that will use __setattr__,
+        # which itself relies on `_query_params` being defined
+        self.__dict__["_query_params"] = {}
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._query_params.keys())
+
+    def __getitem__(self, key: str) -> str:
+        try:
+            value = self._query_params[key]
+            if isinstance(value, list):
+                if len(value) == 0:
+                    return ""
+                else:
+                    return value[-1]
+            return value
+        except:
+            raise KeyError(_missing_key_error_message(key))
+
+    def __setitem__(self, key: str, value: Union[str, List[Any]]) -> None:
+        # Handle non-string inputs like integers or lists by converting them to
+        # strings for consistent query parameter handling.
+        if isinstance(value, list):
+            self._query_params[key] = [str(item) for item in value]
+        else:
+            self._query_params[key] = str(value)
+        self._send_query_param_msg()
+
+    def get_all(self, key: str) -> List[str]:
+        try:
+            if key not in self._query_params:
+                return []
+            query_params = self._query_params[key]
+            return query_params if isinstance(query_params, list) else [query_params]
+        except KeyError:
+            raise KeyError(_missing_key_error_message(key))
+
+    def __getattr__(self, key: str) -> str:
+        try:
+            return self.__getitem__(key)
+        except KeyError:
+            raise AttributeError(_missing_key_error_message(key))
+
+    def __setattr__(self, key: str, value: str) -> None:
+        try:
+            self.__setitem__(key, value)
+        except KeyError:
+            raise AttributeError(_missing_key_error_message(key))
+
+    def __delattr__(self, key: str) -> None:
+        try:
+            self.__delitem__(key)
+        except KeyError:
+            raise AttributeError(_missing_key_error_message(key))
```

Comment:
This works but is weird: `__getitem__` is defined so that you can write `self[key]` instead of `self.__getitem__(key)`. Similar comments for `set`/`del`